src/engine/sparqlExpressions/ConditionalExpressions.cpp

```diff
@@ -0,0 +1,110 @@
+// Copyright 2023, University of Freiburg,
+// Chair of Algorithms and Data Structures
+// Authors: Johannes Kalmbach <kalmbacj@cs.uni-freiburg.de>
+
+#include "NaryExpression.h"
+#include "NaryExpressionImpl.h"
+
+namespace sparqlExpression {
+namespace detail::conditional_expressions {
+using namespace sparqlExpression::detail;
+[[maybe_unused]] auto ifImpl = [](EffectiveBooleanValueGetter::Result condition,
+                                  auto&& i, auto&& e) -> IdOrString {
+  if (condition == EffectiveBooleanValueGetter::Result::True) {
+    return AD_FWD(i);
+  } else {
+    return AD_FWD(e);
+  }
+};
+NARY_EXPRESSION(IfExpression, 3,
+                FV<decltype(ifImpl), EffectiveBooleanValueGetter,
+                   ActualValueGetter, ActualValueGetter>);
+}  // namespace detail::conditional_expressions
+
+// ____________________________________________________
+class CoalesceExpression : public SparqlExpression {
+  std::vector<SparqlExpression::Ptr> children_;
+
+ public:
+  explicit CoalesceExpression(std::vector<SparqlExpression::Ptr> children)
+      : children_{std::move(children)} {};
+  ExpressionResult evaluate(EvaluationContext* context) const override {
+    std::vector<ExpressionResult> childResults;
+    std::ranges::for_each(children_, [&](const auto& child) {
```

Comment:
Why isn't the algorithm in the following simply as follows: evaluate one child after the other until the result is undefined, then return that as a result, UNDEF otherwise.

Reply:
Well, some of the children return vectors of elements.  So your algorithm would have to work basically on rows of (one entry per vector). And this evaluation would be inefficient.

What we can do and probably should (also for the other expressions) is to not evaluate the complete children, but for examples always chunks of N. Then we have both the efficient evaluation of the single children as well as bounded space consumption. Then we can do something like "If in the current block everything is already defined, then we can skip it entirely".